Added spec changes for Image Extension (#267)
buildpacks/spec: buildpack.md

If a buildpack order entry within a group has the parameter `optional = true`, then a copy of the group without the entry MUST be repeated after the original group.

#### Image Extensions

Image extensions participate in the buildpack [detection](#detector) process, with the same `UID`, `GID`, and interface for `/bin/detect`. However:

> - Detection is optional for extensions, and they are assumed to pass detection when it is not present. A `/bin/detect` that exits with a 0 exit code passes detection, and fails otherwise.

Reviewer: ```suggestion
- Detection is optional for extensions, and they are assumed to pass detection when `/bin/detect` is not present. A `/bin/detect` that exits with a 0 exit code passes detection, and fails otherwise.
```

- Extensions MUST only output `provides` entries to the build plan. They MUST NOT output `requires`.

> - Extensions MUST all precede buildpacks in [order](#ordertoml-toml) definitions.

Reviewer: https://github.com/buildpacks/samples/pull/113/files#r726206827 talks about creating a separate `[[order-ext]]` in builder.toml - would that translate into a separate file?

After analysis, the lifecycle MUST proceed to the build phase.

## Phase #3: Image Extension

### Purpose

The purpose of the image extension phase is to apply dynamically-generated or static build-time and runtime modules that mutate the pre-build base image.

During the image extension phase, an image extension might:

> 1. Execute a pre-defined `Dockerfile` to install files in specific directories that require privilaged access.

Reviewer: ```suggestion
1. Execute a pre-defined `Dockerfile` to install files in specific directories that require privileged access.
```

1. Dyanmically generate a `Dockerfile` based on the phase inputs and apply it.
1. Execute a pre-defined `Dockerfile` to install system packages provided as input to the phase.

### Process

**GIVEN:**

> - The final ordered group of builimage extensions determined during the detection phase,

Reviewer 2: ```suggestion
- The final ordered group of image extensions determined during the detection phase,
```

- A directory containing application source code,

> - A `<output>` directory used to store generated artifacts

Reviewer: ```suggestion
- An `<output>` directory used to store generated artifacts
```

- A shell, if needed,

For each image extension in the group in order, the lifecycle MUST execute `/bin/build`.

1. **If** the exit status of `/bin/build` is non-zero, \
   **Then** the lifecycle MUST fail the phase.

2. **If** the exit status of `/bin/build` is zero,
   1. **If** there are additional image extensions in the group, \
      **Then** the lifecycle MUST proceed to the next image extension's `/bin/build`.

   2. **If** there are no additional image extension in the group, \
      **Then** the lifecycle MUST proceed to the build phase.

For each `/bin/build` executable in each image extension, the lifecycle:

- MUST provide path arguments to `/bin/build` as described in the [Image Extension Interface](image-extension.md) section.
- MUST configure the build environment as described in the [Environment](#environment) section.
- MUST provide all `<plan>` entries that were required by any buildpack in the group during the detection phase with names matching the names that the buildpack provided.

Correspondingly, each `/bin/build` executable:

- MAY read from the `<app>` directory.

> - MUST NOT write to the `<app>` directory.

Reviewer 2: ```suggestion
- MUST NOT write to the `<app>` directory.
```

Is this one of those places where we hope they won't but can't do anything if they do?

Author: @jabrown85 i think it depends on the platform. Like, if it's running the extension in it's own container, it might be able to exclude `<app>` dir changes from it's snapshot. But if it's all running on the same filesystem/volume it won't be able to.

This isn't actually a suggested change right?

Reviewer 2: Yeah, not a suggested change. I must have clicked the wrong button :)

- MAY read the build environment as described in the [Environment](#environment) section.
- MAY read the Buildpack Plan.
- MAY log output from the build process to `stdout`.
- MAY emit error, warning, or debug messages to `stderr`.

> - SHOULD write BOM (Bill-of-Materials) entries to `<output>/launch.toml` describing any contributions to the app image.

Reviewer: ```suggestion
- SHOULD write SBOM (Software-Bill-of-Materials) entries as described in the [Software-Bill-of-Materials](#software-bill-of-materials) section describing any contributions to the build environment or app image.
```

> - MAY write key-value pairs to `<output>/launch.toml` that are provided as build args to run.Dockerfile or Dockerfile

Reviewer 2: This is the first time `run.Dockerfile` is mentioned, should we omit this specific file name here in favor of keeping that in the image-extensions.md? Just refer to Dockerfile here in a generic sense maybe...

> - SHOULD write build BOM entries to `<output>/build.toml` describing any contributions to the build environment.

Reviewer: ```suggestion
- SHOULD write build SBOM (Software-Bill-of-Materials) files as described in the [Software-Bill-of-Materials](#software-bill-of-materials) section describing any contributions to the build environment.
```

- MAY modify or delete any existing `<output>` directories.
- MAY create new `<output>` directories.

> - MAY name any new `<output>` directories without restrictions except those imposed by the filesystem and the ones noted below.

Reviewer: I'm not quite sure what to make of these bullets. I think they could be removed (when talking about `<layers>` they make sense but we don't really care about the output directory for extensions, only the files contained there).
